#166214112 Implement Email And In-app Notification

henryade opened the pull request:
#### What does this PR do?
adds an email and in-app notification 
#### Description of Task to be completed?
-  create a notification model
- create a user-notification model
- create a notification helper class for all notification process
- optional email notification feature
- use socket.io for in-app notifications
#### How should this be manually tested?
#### To test the in-app notification with socket io
- create an HTML file and copy this into it
```
<script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.2.0/socket.io.js"></script>
    <script>
      const socket = io.connect('http://localhost:PORT');
        socket.emit('showNotifications', { id: 9 });
        socket.on('showNotifications', function(data){
          console.log("data");
          console.log(data);
        });
    </script>
```
- change the PORT to the port specified in your env
- make sure your database is seeded
- run your server before you open the HTML file in your browser. Open the console to see data received
- if you don't see it at first, refresh the browser

#### To test the email notification
FOR NEW BOOK BY FAVOURITE AUTHOR NOTIFICATION
- change either the 3rd, 8th or 9th users email in the database to a valid email.
- By doing the above step means you favorited the first author in the database
- run `npm run db:rollback` to seed db again then create a book with the first author in the db.
- check your mail for notification

FOR REMINDER OF BORROWED BOOK DUE DATE NOTIFICATION
- change either the 3rd user email in the database to a valid email.
- create a token that expires in 2days and change the token in the lending history seeders file
- run `npm run db:rollback` to seed database again then sign in with an admin user.
- check your mail for notification

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166214112](https://www.pivotaltracker.com/story/show/166214112)
#### Screenshots (if appropriate)
IN-APP NOTIFICATION:
![image](https://user-images.githubusercontent.com/47576037/59888892-aa800a80-93c1-11e9-918d-c5d645b57216.png)

NEW BOOK EMAIL:
![image](https://user-images.githubusercontent.com/47576037/59888679-c8993b00-93c0-11e9-8248-190fe20d94e0.png)

REMAINDER FOR BORROWED BOOK EMAIL:
![image](https://user-images.githubusercontent.com/47576037/59888649-a1426e00-93c0-11e9-86c1-f7cd6210c26e.png)

#### Questions:
N/A

Lundii reviewed (commented):
For the in app notification test, What should be the name of the html file we are to create and where should we create and place it.

henryade commented:
> For the in app notification test, What should be the name of the html file we are to create and where should we create and place it.

create it anywhere and give it any name

Lundii commented:
> > For the in app notification test, What should be the name of the html file we are to create and where should we create and place it.
> 
> create it anywhere and give it any name

Okay. Will look into it. I wanna ask though, is it only a patron that receives notification?

Lundii commented:
> > > For the in app notification test, What should be the name of the html file we are to create and where should we create and place it.
> > 
> > 
> > create it anywhere and give it any name
> 
> Okay. Will look into it. I wanna ask though, is it only a patron that receives notification?

So, I tested the PR on my local (the in app notification part) and it was working like you said. But I'm yet to understand whats happening. The messages sent to the console, where are they coming from. How can I get an in app notification for different users. Just give a break down on how the in app notification works. You can send me a voice note on whats app or write it here so others can read it. Thanks

Ukhu reviewed (commented):
Rebase branch so I can approve

ad3ola committed:
166214112-story(notification): implement email and in-app notification
create notfication model
create userNotification model
create notification helper class for all notification process
use socket.io for in-app notifications
[finishes #166214112]
